crates/sound/src/audio_backends/alsa.rs

```diff
@@ -0,0 +1,550 @@
+                .unwrap_or_default()
+        })?;
+        if written == 0 {
+            break;
+        };
+    }
+
+    match p.state() {
+        State::Suspended | State::Running => Ok(false), // All fine
+        State::Prepared => {
+            //p.start()?;
+            Ok(false)
+        }
+        State::XRun => Ok(true), // Recover from this in next round
+        n => panic!("Unexpected pcm state {:?}", n),
+    }
+}
+
+fn alsa_worker(
+    pcm: Arc<Mutex<PCM>>,
+    streams: Arc<RwLock<Vec<Stream>>>,
+    receiver: &Receiver<bool>,
+    stream_id: usize,
+) -> AResult<()> {
+    loop {
+        let Ok(do_write) = receiver.recv() else {
+            return Ok(());
+        };
+        if do_write {
+            loop {
+                if matches!(receiver.try_recv(), Ok(false)) {
+                    break;
+                }
+
+                let mut fds = {
+                    let lck = pcm.lock().unwrap();
+                    if matches!(lck.state(), State::Running | State::Prepared | State::XRun) {
+                        let mut mmap = lck.direct_mmap_playback::<u8>().ok();
+
+                        if let Some(ref mut mmap) = mmap {
+                            if write_samples_direct(
+                                &lck,
+                                &mut streams.write().unwrap()[stream_id],
+                                mmap,
+                            )? {
+                                continue;
+                            }
+                        } else {
+                            let mut io = lck.io_bytes();
+                            // Direct mode unavailable, use alsa-lib's mmap emulation instead
+                            if write_samples_io(
+                                &lck,
+                                &mut streams.write().unwrap()[stream_id],
+                                &mut io,
+                            )? {
+                                continue;
+                            }
+                        }
+                        lck.get()?
+                    } else {
+                        drop(lck);
+                        sleep(Duration::from_millis(500));
+                        continue;
+                    }
+                };
+                // Nothing to do, let's sleep until woken up by the kernel.
+                alsa::poll::poll(&mut fds, 100)?;
+            }
+        }
+    }
+}
+
+impl AlsaBackend {
+    pub fn new(streams: Arc<RwLock<Vec<Stream>>>) -> Self {
+        let (sender, receiver) = channel();
+        let sender = Arc::new(Mutex::new(sender));
+
+        thread::spawn(move || {
+            if let Err(err) = Self::run(streams, receiver) {
+                log::error!("Main thread exited with error: {}", err);
+            }
+        });
+
+        Self { sender }
+    }
+
+    fn run(
+        streams: Arc<RwLock<Vec<Stream>>>,
+        receiver: Receiver<AlsaAction>,
+    ) -> std::result::Result<(), Box<dyn std::error::Error>> {
```

Comment:
```suggestion
    ) -> Result<(), Box<dyn std::error::Error>> {
```
As you have a `AResult` / `CrateResult`  already.

Reply:
Fixed!